### ISSUE TITLE
Use generics in sorting algorithms

### ISSUE THREAD

**Issue by meetvalani** [TheAlgorithms/Go#419]
**Description**
IMO algorithm should be working on data type `float64` instead of `int`.

For enhancement:
- All the algorithms should be converted to use `float64` as an arguments. Followings are the possible enhancements.
    - User must handle the type conversion of `int` to `float64`.
    - Should provide both type where `int` internally uses `float64` related function.

**Comment by tjgurwara99** [comment]
In case someone wants to experiment with generics [here](https://go2goplay.golang.org) is the link to the GoTip playground to test out new features 😄

**Comment by tjgurwara99** [comment]
closed in #483

**Comment by tjgurwara99** [comment]
Closed by mistake, we still have three sorting algorithms that need to be modified to use the generic. Heap sort, radix sort and pigeonhole sort all require some refactoring for us to be able to convert them to generic sorting algorithms.

**Comment by jaiswalpuru** [comment]
Can I pick this up ?

**Comment by mohamedspicer** [comment]
@tjgurwara99 can you assign this to me

**Comment by tjgurwara99** [comment]
You don't need an assignment of issues. Feel free to open a PR and we would review it 😄

**Comment by VUHAILAM** [comment]
I researched radix sort can apply for string: [](https://www.quora.com/How-can-I-sort-strings-using-Radix-Sort). But our radix sort is only for Integer. Should we implement another radix sort can apply both??

**Comment by tjgurwara99** [comment]
> But our radix sort is only for Integer.

Yes, therefore we have this issue to convert it to a generic one.

> Should we implement another radix sort can apply both??

Radix sort is **Radix sort**, therefore there is really no difference in them unless you take a radically novel approach to writing this algorithm. Therefore, my suggestion is to modify the existing one, **unless** your implementation is resulting in a difference in complexity (either space or time)

**Comment by badsector998** [comment]
> > But our radix sort is only for Integer.
> 
> Yes, therefore we have this issue to convert it to a generic one.
> 

I'd like to collaborate more on this. By converting integer to generic, do you mean we want to apply radix sort that can accept other data type beside integer?

**Comment by siriak** [comment]
Yes. As it was already mentioned, it can be applied to strings. Integers are numbers base 10 and this sort can apply to numbers with any base, but representing them is problematic. We could use strings for that though like we use strings to represent hex numbers.

**Comment by badsector998** [comment]
I have been thinking, since it can sort string and integer or any base number type, can we try to sort them by their binary representation? If we sort them by keeping the native data type I think the implementation wouldn't be generic. I am planning to dig deep on this and may be trying some implementation too by this weekend.

**Comment by LaviVasudevan** [comment]
Hi, I’m participating in Hacktoberfest. I’d like to work on converting one of the remaining sorting algorithms (eg HeapSort) to generics. Please let me know if that’s okay, or if you’d prefer me to handle a different algorithm.

**Comment by LeonxLJX** [comment]
I'd like to take this one (`Use generics in sorting algorithms`). I'll look into the root cause and follow up with a PR. (claiming via @LeonxLJX)
